chain/index/api.go

```diff
@@ -0,0 +1,167 @@
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	// Canonical chain has a null round at the epoch -> return if index is empty otherwise validate
+	if expectedTs.Height() != epoch { // Canonical chain has a null round at the epoch
+		if isIndexEmpty {
+			return nil, nil
+		}
+		return si.validateNullRound(ctx, epoch)
+	}
+
+	// if the index is empty -> short-circuit and simply backfill if applicable
+	if isIndexEmpty {
+		return si.backfillMissingTipset(ctx, expectedTs, backfill)
+	}
+
+	revertedCount, nonRevertedCount, err := si.getTipsetCountsAtHeight(ctx, epoch)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return si.backfillMissingTipset(ctx, expectedTs, backfill)
+		}
+		return nil, xerrors.Errorf("failed to get tipset counts at height: %w", err)
+	}
+	if revertedCount == 0 && nonRevertedCount == 0 {
+		return si.backfillMissingTipset(ctx, expectedTs, backfill)
+	}
+
+	if revertedCount > 0 && nonRevertedCount == 0 {
+		return nil, xerrors.Errorf("index corruption: height %d only has reverted tipsets", epoch)
+	}
+
+	if nonRevertedCount > 1 {
+		return nil, xerrors.Errorf("index corruption: height %d has multiple non-reverted tipsets", epoch)
+	}
+
+	// fetch the non-reverted tipset at this epoch
+	var indexedTsKeyCidBytes []byte
+	err = si.stmts.getNonRevertedTipsetAtHeightStmt.QueryRowContext(ctx, epoch).Scan(&indexedTsKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get non-reverted tipset at height: %w", err)
+	}
+	indexedTsKeyCid, err := cid.Cast(indexedTsKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to cast tipset key cid: %w", err)
+	}
+
+	if indexedTsKeyCid != expectedTsKeyCid {
+		return nil, xerrors.Errorf("index corruption: non-reverted tipset at height %d has key %s, but canonical chain has %s", epoch, indexedTsKeyCid, expectedTsKeyCid)
+	}
+
+	return &types.IndexValidation{
+		TipsetKey: expectedTs.Key().String(),
+		Height:    uint64(expectedTs.Height()),
+		// TODO Other fields
+	}, nil
+}
+
+func (si *SqliteIndexer) backfillMissingTipset(ctx context.Context, ts *types.TipSet, backfill bool) (*types.IndexValidation, error) {
+	if !backfill {
+		return nil, xerrors.Errorf("missing tipset at height %d in the chain index, set backfill to true to backfill", ts.Height())
+	}
+	// backfill the tipset in the Index
+	parentTs, err := si.cs.GetTipSetFromKey(ctx, ts.Parents())
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get parent tipset: %w", err)
+	}
+
+	err = withTx(ctx, si.db, func(tx *sql.Tx) error {
+		if err := si.indexTipsetWithParentEvents(ctx, tx, parentTs, ts); err != nil {
+			return xerrors.Errorf("error indexing tipset: %w", err)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, xerrors.Errorf("error applying tipset: %w", err)
+	}
+
+	return &types.IndexValidation{
+		TipsetKey:  ts.Key().String(),
+		Height:     uint64(ts.Height()),
+		Backfilled: true,
```

Comment:
todo: other fields

Reply:
@rvagg What other fields do you have in mind here ?

Reply:
Done.